pkg/apis/pipeline/v1alpha1/pipeline_validation.go

```diff
@@ -204,15 +204,97 @@ func (ps *PipelineSpec) Validate(ctx context.Context) *apis.FieldError {
 		return err
 	}
 
+	if err := validatePipelineParameterVariables(ps.Finally, ps.Params); err != nil {
+		return err
+	}
+
 	// Validate the pipeline's workspaces.
-	if err := validatePipelineWorkspaces(ps.Workspaces, ps.Tasks); err != nil {
+	if err := validatePipelineWorkspaces(ps.Workspaces, ps.Tasks, ps.Finally); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func validatePipelineWorkspaces(wss []WorkspacePipelineDeclaration, pts []PipelineTask) *apis.FieldError {
+func validateFinalTasks(ps *PipelineSpec) *apis.FieldError {
+	if len(ps.Finally) != 0 && len(ps.Tasks) == 0 {
+		return apis.ErrInvalidValue(fmt.Sprintf("spec.tasks is empty but spec.finally has %d tasks", len(ps.Finally)), "spec.finally")
+	}
+
+	if ps.Finally != nil && len(ps.Finally) == 0 {
+		return apis.ErrInvalidValue("spec.finally is empty", "spec.finally")
+	}
+
+	for _, f := range ps.Finally {
+		if len(f.RunAfter) != 0 {
+			return apis.ErrInvalidValue(fmt.Sprintf("no runAfter allowed under spec.finally, final task %s has runAfter specified", f.Name), "spec.finally")
+		}
+		if len(f.Conditions) != 0 {
+			return apis.ErrInvalidValue(fmt.Sprintf("no conditions allowed under spec.finally, final task %s has conditions specified", f.Name), "spec.finally")
+		}
+	}
+	return nil
+}
```

Comment:
do we want something in here to look for "results" use as well or is that covered in other code?

Reply:
`results` are part of tasks, we are validating `PipelineTask` here. Also, no validation done on whether a final task is looking for result. In case a final task references a task result in a param e.g. `$(tasks.count-comments-after.results.count)`, that param value is not resolved and results in[ not found](https://github.com/tektoncd/pipeline/pull/2437/files#diff-db9ac78ce08367ea2b7db8c1412049b6R691) during the pipeline execution.

I think its a good point to catch it and fail at the validation, will add it.

Reply:
Added, https://github.com/tektoncd/pipeline/pull/2650/files#diff-5b096c7641765902d79ee2d5e5f7d213R465